rfd/0091-session-streaming.md

```diff
@@ -0,0 +1,238 @@
+
+## Why
+
+There are several reasons why we'd like to make this change.
+
+### Removal of Legacy Code
+
+The `lib/events` package has a long history of [issues](https://github.com/gravitational/teleport/issues?q=is%3Aopen+is%3Aissue+label%3Aaudit-log)
+and dead code that
+[should](https://github.com/gravitational/teleport/pull/13217) 
+[be](https://github.com/gravitational/teleport/pull/12395) 
+[removed](https://github.com/gravitational/teleport/pull/11380).
+
+While we've succeeded in removing a large amount of old code, there is still
+a non-trivial amount of code left behind due to the legacy playback format
+that can be removed once we support the newer streaming API.
+
+Reducing the surface area of the packages related to the audit log and session
+recordings will make it easier to understand and modify as we continue to work
+on new efforts such as an encrypted audit log.
+
+### User Experience
+
+The required conversion to the old playback format results in an experience that
+is suboptimal for users. Since this process requires the entire session be 
+downloaded and converted before playback can begin, users experience significant
+latency when attempting to play back a large session.
+
+A prototype of streaming playback with a large session has shown a > 10x 
+improvement in "time to first byte" of playback (11s to < 1s).
+
+In addition to latency, the current approach requires the entire session to be
+loaded into memory. This is inefficient, and consuming large amounts of system 
+memory can impact the overall responsiveness of the the system.
+
+In the web UI, this can also result in failure to play the session at all,
+as observed in [#10578](https://github.com/gravitational/teleport/issues/10578).
+
+### Consistency with Desktop Access
+
+When session recording and playback for desktop access was implemented
+(see [RFD 48](./0048-desktop-access-session-recording.md) for details),
+we needed to build a new graphical session player, so there was no reason
+to convert to the legacy format in order to reuse the text-based SSH player.
+
+As a result, desktop access has supported streaming playback since its 
+inception. Implementing this RFD for other types of sessions will ensure
+greater consistency across the codebase, and will address the long-standing
+request to implement seek functionality for desktop playback.
+
+## Details
+
+In all cases, we will be leveraging the `StreamSessionEvents` API to do session 
+playback. This API presents the following Go interface:
+
+```
+func StreamSessionEvents(
+    ctx context.Context, 
+    sessionID session.ID, 
+    startIndex int64,
+  ) (chan apievents.AuditEvent, chan error) {
+```
+
+A gRPC wrapper is also available:
+
+```
+  rpc StreamSessionEvents(StreamSessionEventsRequest) returns (stream events.OneOf);
+```
+
+Given a session ID and a start index, a channel of `AuditEvent`s is returned.
+Consumers can control the speed of playback by limiting how fast they consume
```

Comment:
👍  Playback speed control is something I see requested a fair bit, so this would be neat.